kuzzleio/kuzzle
# Framework: allow the execution of code only once

## What does this PR do ?

Add an install method available for Kuzzle as a Framework allowing developers to run a piece of code only once during the lifetime of the app. This can be useful when considering db migrations or setup with external services and many more use cases...

### How should this be manually tested?

- Initialize a kuzzle app `kourou app:scaffold playground`
- Open the file app.ts
- Between `const app = new Backend('playground')` and `app.start()` write `app.install('id', () => { console.log('I will be print only once') });`
- Run kuzzle `docker-compose up`
- Watch the logs and see your message 

### Other changes

- Classes inside lib/core/backend/backend.ts and their tests (test/core/backend/Backend.test.js) extracted in several classes (Review tip: without changing code inside these classes)
  - backendConfig
  - backendController
  - backendHook
  - backendPipe
  - backendStorage
  - backendVault
  - internalLogger
- Deleted type export internalLogger since the class is now exported directly
- Doc for install method
- Unit tests added (Review tip: in Backend.test.js and kuzzle.test.js)
- Added a collection %kuzzle.installations in default.config.js
- Fix versioning in the RequestInput doc
